src/steps/pre-query/parse-request-primary.js

```diff
@@ -22,14 +22,24 @@ export default function(data, parseAsLinkage) {
     }
 
     catch (error) {
-      const title = "The resources you provided could not be parsed.";
-      const details = `The precise error was: "${error.message}".`;
-      reject(new APIError(400, undefined, title, details));
+      if (error instanceof APIError) {
+        reject(error);
+      }
+
+      else {
+        const title = "The resources you provided could not be parsed.";
+        const details = `The precise error was: "${error.message}".`;
+        reject(new APIError(400, undefined, title, details));
+      }
     }
   });
 }
 
-function relationshipObjectFromJSON(json) {
+function relationshipObjectFromJSON(json, relationship) {
+  if (typeof json.data === "undefined") {
+    throw new APIError(400, undefined, `No data key was found for the '${relationship}' relationship.`);
```

Comment:
Overall, this PR looks great. I'd make one small tweak, though, namely: I'd throw a more generic error from `relationshipObjectFromJSON` and then augment and rethrow that error in `resourceFromJSON`. Like so:

```js
function relationshipObjectFromJSON(json) {
  if (typeof json.data === "undefined") {
    throw new APIError(400, undefined, "Missing relationship linkage.");
  }

  //  blah blah
}

// then...
function resourceObjectFromJSON(json) {
  // blah blah

  // build RelationshipObjects
  let key;
  try {
    for(key in relationships) {
      relationships[key] = relationshipObjectFromJSON(relationships[key]);
    }
  }
  catch(e) {
    e.details =  `No data key was found for the '${key}' relationship.`;
    throw e;
  }

  // blah blah
}
```

I prefer this because adding the `relationship` argument to `relationshipObjectFromJSON` just for use in an error message bothered me a bit. It would suggest that the creation of the relationship object is dependent on knowing the relationship's name, when that's not really true or is only true in a trivial sense. Also, this has the side benefit of making the APIError's `title` into something constant (the "Missing Relationship Linkage" message) while making `details` vary per instance of the problem (the "No data key was found..." message), which is how title and details are specced to work.